c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -399,103 +399,76 @@ cudaError_t cudaMallocMaybeCapturing(void** p, size_t size) {
-      : m_max_split_size(std::numeric_limits<size_t>::max()),
-        m_roundup_power2_divisions(0),
-        m_garbage_collection_threshold(0) {}
-  size_t m_max_split_size;
-  size_t m_roundup_power2_divisions;
-  double m_garbage_collection_threshold;
-
-  void parseArgs() {
-    const char* val = getenv("PYTORCH_CUDA_ALLOC_CONF");
-    if (val != NULL) {
-      const std::string config(val);
-
-      std::regex exp("[\\s,]+");
-      std::sregex_token_iterator it(config.begin(), config.end(), exp, -1);
-      std::sregex_token_iterator end;
-      std::vector<std::string> options(it, end);
-
-      for (auto option : options) {
-        std::regex exp2("[:]+");
-        std::sregex_token_iterator it2(option.begin(), option.end(), exp2, -1);
-        std::sregex_token_iterator end2;
-        std::vector<std::string> kv(it2, end2);
-        if (kv.size() >= 2) {
-          /* Maximum split size in MB.  Limited to large size blocks */
-          if (kv[0].compare("max_split_size_mb") == 0) {
-            size_t val2 = stoi(kv[1]);
-            TORCH_CHECK(
-                val2 > kLargeBuffer / (1024 * 1024),
-                "CachingAllocator option max_split_size_mb too small, must be > ",
-                kLargeBuffer / (1024 * 1024),
-                "");
-            val2 = std::max(val2, kLargeBuffer / (1024 * 1024));
-            val2 = std::min(
-                val2, (std::numeric_limits<size_t>::max() / (1024 * 1024)));
-            m_max_split_size = val2 * 1024 * 1024;
-          } else if (kv[0].compare("roundup_power2_divisions") == 0) {
-            size_t val2 = stoi(kv[1]);
-            TORCH_CHECK(
-                llvm::isPowerOf2_64(val2),
-                "For roundups, the divisons has to be power of 2 ",
-                "");
-            m_roundup_power2_divisions = val2;
-          } else if (kv[0].compare("garbage_collection_threshold") == 0) {
-            /*
-             * Perform garbage collection of GPU memory blocks to avoid
-             * triggering expensive sync-and-reclaim-all operation. Upon setting
-             * the threshold (e.g., 0.8), the allocator will start reclaiming
-             * blocks if GPU memory capacity usage exceeds the threshold (i.e.,
-             * 80% of total memory).
-             * Values 0.0 and 1.0 are not allowed as they are less meaningful.
-             */
-            double val2 = stod(kv[1]);
-            TORCH_CHECK(
-                val2 > 0,
-                "garbage_collect_threshold too small, set it 0.0~1.0",
-                "");
-            TORCH_CHECK(
-                val2 < 1.0,
-                "garbage_collect_threshold too big, set it 0.0~1.0",
-                "");
-            m_garbage_collection_threshold = val2;
-          } else {
-            TORCH_CHECK(false, "Unrecognized CachingAllocator option: ", kv[0]);
-          }
-        }
+void CachingAllocatorConfig::parseArgs(const char* env) {
+  // If empty, set the default values
```

Comment:
If I only set one config option here, but another option was set to a different value before, then we inherit the old value with this logic. I would expect the way it would work is all options that are not specified in the configuration string to go back to the default value (i.e. universally set everything to defaults, then modify the specific options, if any).

Reply:
yes, this is the current behavior, setting not specified, goes back to default

Reply:
What happens if I first set it to "max_split_size_mb=4", then set it to "garbage_collection_threshold=0.8". My reading of this code is that`max_split_size_mb` will still be set to `4` but I would expect it set to `size_t::max()`